### PR TITLE
Limit open journals to avoid running out of file descriptors

### LINKED ISSUE #307: Wash panics with too many files opened by activity journal
Running something like
`find aws/prof/resources/ec2/instances -meta .Tags[?] .Key department | xargs -n1 meta | jq '.Tags[].Key == "department"'` caused
```
panic: journal entry open /Users/michaelsmith/Library/Caches/wash/activity/16239-sh.log: too many open files was not a Logger

goroutine 34 [running]:
github.com/puppetlabs/wash/activity.closeJournal(0xc006de8440, 0xa, 0x54cb320, 0xc003ba0e10)
	/Users/michaelsmith/.go/src/github.com/puppetlabs/wash/activity/core.go:87 +0x28e
github.com/ekinanp/go-cache.(*cache).DeleteExpired(0xc00010d600)
	/Users/michaelsmith/.go/pkg/mod/github.com/ekinanp/go-cache@v2.1.0+incompatible/cache.go:946 +0x419
github.com/ekinanp/go-cache.(*janitor).Run(0xc0000f3a40, 0xc00010d600)
	/Users/michaelsmith/.go/pkg/mod/github.com/ekinanp/go-cache@v2.1.0+incompatible/cache.go:1081 +0x60
created by github.com/ekinanp/go-cache.runJanitor
	/Users/michaelsmith/.go/pkg/mod/github.com/ekinanp/go-cache@v2.1.0+incompatible/cache.go:1099 +0xa9

zsh: error on TTY read: Input/output error
```

Looks like there are too many open files. When this happens we should start closing older journals.

### PR DESCRIPTION
**Pull request by MikaelSmith**
Fixes #307.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

### CODE REVIEW

File: datastore/cache.go [hunk 1]
```diff
@@ -58,11 +62,21 @@ func (cache *MemCache) lockForKey(category, key string) *locksutil.LockEntry {
...
+	cache.limit = n
+	return cache
+}
+
 // GetOrUpdate attempts to retrieve the value stored at the given key.
 // If the value does not exist, then it generates the value using
 // the generateValue function and stores it with the specified ttl.
 // If resetTTLOnHit is true, will reset the cache expiration for the entry.
 func (cache *MemCache) GetOrUpdate(category, key string, ttl time.Duration, resetTTLOnHit bool, generateValue func() (interface{}, error)) (interface{}, error) {
+	cache.mux.RLock()
```
[the remark this answers is not in the document]
Comment (MikaelSmith) [reply datastore/cache.go]: I was seeing panics from concurrent map read/write, as I noted with the mux declaration.

**Comment by MikaelSmith** [comment]
Added the builder pattern.

File: datastore/cache.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ekinanp) [reply datastore/cache.go]: Which map? Where does the panic happen? Those aren't clear from the current comments

[the remark this answers is not in the document]
Comment (MikaelSmith) [reply datastore/cache.go]: go-cache operations are not thread safe and are basically fancy map manipulation. I can update the comment later today to mention the cache’s internal map.

[the remark this answers is not in the document]
Comment (MikaelSmith) [reply datastore/cache.go]: Done.

**Pull request merged**